### ISSUE TITLE
Provide an example with Bacon number that compiles

### ISSUE THREAD

**Issue by akrzemi1** [stdgraph/graph-v2#85]
First, thank you for this great work on the world-class graph library!

[P1709r4](https://www.open-std.org/jtc1/sc22/wg21/docs/papers/2023/p1709r4.pdf) gives a very promising example of with Bacon number that shows:
* How I can use my own data structure for representing a graph
* That this code can be very brief.

But it does not really compile, and trying to fix it does not work for me. I tried applying fixes, but it still won't compile. In fact, even the following code does not compile:

```c++
using G = std::vector<std::vector<int>>;

auto target_id(const G& g, std::graph::edge_reference_t<G> uv)
{ return uv; }

static_assert(std::graph::basic_adjacency_list<G>, "non-graph");
```

And this is because there is no way for me to create an ADL-discoverable funciton `target_id` because `G` -- as defined in [P1709r4](https://www.open-std.org/jtc1/sc22/wg21/docs/papers/2023/p1709r4.pdf)  -- is composed only from `int` and declarations from the namespace `std`, where I am not allowed to put overloads.

Also, the concept seems strange:

```c++
template <class G, class E>
concept targeted_edge = requires(G&& g, edge_reference_t<G> uv) {
  target_id(g, uv);
  target(g, uv);
};
```

It requires that `target_id(g, uv)` exists, but does not require that it returns anything.

But if I nonetheless define `target_id` inside namespace `std` (which is UB), the following fails to compile:

```c++
G costar_adjacency_list{
  {1, 5, 6}, {7, 10, 0, 5, 12}, {4, 3, 11}, {2, 11}, {8, 9, 2, 12}, {0, 1}, {7, 0},
  {6, 1, 10}, {4, 9}, {4, 8}, {7, 1}, {2, 3}, {1, 4}
};

std::graph::views::sourced_edges_breadth_first_search(costar_adjacency_list, 1);
```

I get an error:

```
graph-v2/include/graph/views/breadth_first_search.hpp:73:29: error: no matching function for call to object of type 'const _Vertices::_Cpo'
    if (seed < ranges::size(vertices(graph_)) && !ranges::empty(edges(graph_, seed))) {
                            ^~~~~~~~
```

Is it because I have to provide an overload for `vertices(G)`? But do I, given that `G` itself is a list of vertices? 

One compiling example would help answer all these questions.

**Comment by akrzemi1** [comment]
Ping?

**Comment by akrzemi1** [comment]
I tried it myself with the following example. It is no the same as the example from [P1709r4](https://www.open-std.org/jtc1/sc22/wg21/docs/papers/2023/p1709r4.pdf). I had to change the type of vertex from `int` to `ak::Box`, because I needed my namespace to enable ADL. But it still won't compile (GCC 13.2). Can you tell me what I need to do to make it compile?

```c++

#include <iostream>
#include <string>
#include <vector>
#include <graph/views/breadth_first_search.hpp>

std::vector<std::string> actors{
  "Tom Cruise",
  "Kevin Bacon",
  "Hugo Weaving",
  "Carrie-Anne Moss",
  "Natalie Portman",
  "Jack Nicholson",
  "Kelly McGillis",
  "Harrison Ford",
  "Sebastian Stan",
  "Mila Kunis",
  "Michelle Pfeiffer",
  "Keanu Reeves",
  "Julia Roberts",
};

namespace gview = std::graph::views;

namespace ak {

struct Box
{
  int val;
  Box(int i) : val(i) {}
};



  using G = std::vector<std::vector<Box>>;

int target_id(const G&, const Box& uv) noexcept { return uv.val; }

}

std::vector<std::vector<ak::Box>> costars{
    {1, 5, 6},
    {7, 10, 0, 5, 12},
    {4, 3, 11},
    {2, 11},
    {8, 9, 2, 12},
    {0, 1},
    {7, 0},
    {6, 1, 10},
    {4, 9},
    {4, 8},
    {7, 1},
    {2, 3},
    {1, 4}};


int main() {

print_type<std::graph::vertex_id_t<ak::G>>();
  std::vector<int> bacon_number(size(actors));

  for (auto&& [u, v] : gview::edges_breadth_first_search(costars, 1)) {
    bacon_number[v.val] = bacon_number[u] + 1;
  }

  static_assert(std::graph::adjacency_list< ak::G >);
  for (int i = 0; i < size(actors); ++i) {
    std::cout << actors[i] << " has Bacon number " << bacon_number[i] << std::endl;
  }

  return 0;
}

```

**Comment by akrzemi1** [comment]
I also note that there is this Bacon example in `examples`: https://github.com/stdgraph/graph-v2/blob/master/tests/examples_tests.cpp

But it has been commented out with a `#if` as though authors also had problems with making it compile.

**Comment by pratzl** [comment]
I encountered a problem a few weeks ago but haven't gotten back to it - trying to get P1709 complete. I suspect there's an issue with name lookup. I'll work on it this weekend.

**Comment by pratzl** [comment]
I've done some prep work to eliminate potential dependency issues when using tag_invoke for BFS views, and some other issues that might cause problems, to minimize following red herrings. After spending Sunday doing that and sending P1709r5 with the latest updates to the next mailing list, I haven't resolved this issue yet.

**Comment by pratzl** [comment]
I've fixed the problem and the Bacon code in tests/example_tests.cpp is now working. When using vector<vector<int>> to define a graph I found that I had to define a target_id() in std::, which is not a good thing. I've updated the implementation so that it will detect that pattern as well as vector<vector<tuple<int, ...>>> and use the int shown as the target_id. While I show vector, the requirement is that the outer & inner ranges are forward ranges, and int just needs to be integral.
